google-cloud-spanner/src/main/java/com/google/cloud/spanner/AbstractReadContext.java

```diff
@@ -148,27 +247,24 @@ public void onTransactionMetadata(Transaction transaction) {
     @GuardedBy("txnLock")
     private ByteString transactionId;
 
-    MultiUseReadOnlyTransaction(
-        SessionImpl session, TimestampBound bound, SpannerRpc rpc, int defaultPrefetchChunks) {
-      super(session, rpc, defaultPrefetchChunks);
+    MultiUseReadOnlyTransaction(Builder builder) {
+      super(builder);
       checkArgument(
-          bound.getMode() != TimestampBound.Mode.MAX_STALENESS
-              && bound.getMode() != TimestampBound.Mode.MIN_READ_TIMESTAMP,
-          "Bounded staleness mode %s is not supported for multi-use read-only transactions."
-              + " Create a single-use read or read-only transaction instead.",
-          bound.getMode());
-      this.bound = bound;
-    }
-
-    MultiUseReadOnlyTransaction(
-        SessionImpl session,
-        ByteString transactionId,
-        Timestamp timestamp,
-        SpannerRpc rpc,
-        int defaultPrefetchChunks) {
-      super(session, rpc, defaultPrefetchChunks);
-      this.transactionId = transactionId;
-      this.timestamp = timestamp;
+          !(builder.bound != null && builder.transactionId != null)
+              && !(builder.bound == null && builder.transactionId == null),
+          "Either TimestampBound or TransactionId must be specified");
+      if (builder.bound != null) {
+        checkArgument(
+            builder.bound.getMode() != TimestampBound.Mode.MAX_STALENESS
+                && builder.bound.getMode() != TimestampBound.Mode.MIN_READ_TIMESTAMP,
+            "Bounded staleness mode %s is not supported for multi-use read-only transactions."
+                + " Create a single-use read or read-only transaction instead.",
+            builder.bound.getMode());
+        this.bound = builder.bound;
+      } else {
+        this.timestamp = builder.timestamp;
+        this.transactionId = builder.transactionId;
+      }
```

Comment:
I would prefer not to. The reason for introducing this builder pattern in this PR is that this PR requires an additional object to be passed in to the different transaction classes. This would add another parameter to the constructors of these classes, bringing the total number to 6 in this specific case. As a general rule of thumb, a method (or constructor) in Java should not take more than 4 arguments: https://rules.sonarsource.com/java/RSPEC-107